### PR TITLE
[WIP] Implement search in workspace feature

### PR DESCRIPTION
**Pull request by lmcgupe**
Signed-off-by: Guy Perron <guy.perron@ericsson.com>

### CODE REVIEW

File: packages/grep-workspace/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Theia - Grep Workspace Extension
```
Comment (akosyakov) [inline packages/grep-workspace/README.md]: It should not be `grep` extension for sure, it is just implementation details of the search engine, it should be possible to change the implementation without renaming the package.

Comment (akosyakov) [reply packages/grep-workspace/README.md]: @svenefftinge Should not actually the search functionality belong the workspace package?

Comment (lmcgupe) [reply packages/grep-workspace/README.md]: I am in a very early stage of the implementation. We had internal discussions last week about this. Things are not set in stone for the moment. Thanks for the feedback.

Comment (akosyakov) [reply packages/grep-workspace/README.md]: I know :) btw have you looked at the code of VS code? Maybe we can learn something from them.

Comment (akosyakov) [reply packages/grep-workspace/README.md]: e.g. there is the node library for the glob search: https://www.npmjs.com/package/glob
maybe there is similar for regex if we need to support it

Comment (svenefftinge) [reply packages/grep-workspace/README.md]: I think it's fine to have it separated from the workspace extension.

Comment (hexa00) [reply packages/grep-workspace/README.md]: @akosyakov  the idea was to have 1 generic package that you can change the implementation and 1 package that contains the implementation.

I think we could replace grep-workspace to search-workspace as the generic package name ?

Comment (hexa00) [reply packages/grep-workspace/README.md]: @lmcgupe let's rename the implementation 'ripgrep-workspace' and we'll have the general interface 'search-workspace' 

The class names should follow

File: packages/grep-workspace/src/node/grep-process-module.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { ContainerModule, Container } from 'inversify';
+
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-process-module.ts]: empty line

File: packages/grep-workspace/src/node/grep-process-module.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import { ProcessManager } from "@theia/process/lib/node";
+
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-process-module.ts]: empty line

File: packages/grep-workspace/src/node/grep-process-module.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import { ILogger } from '@theia/core/lib/common/logger';
+import { GrepWS, GrepProcessOptions, GrepProcessFactory } from './grep-workspace';
+
+export const grepProcessModule = new ContainerModule(bind => {
+    bind(GrepWS).toSelf().inTransientScope();
+    bind(GrepProcessFactory).toFactory(ctx =>
+        (options: GrepProcessOptions) => {
+            const child = new Container({ defaultScope: 'Singleton' });
+            child.parent = ctx.container;
+
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-process-module.ts]: empty line

File: packages/grep-workspace/src/node/grep-process-module.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+import { ContainerModule, Container } from 'inversify';
+
+import { ProcessManager } from "@theia/process/lib/node";
+
+import { ILogger } from '@theia/core/lib/common/logger';
+import { GrepWS, GrepProcessOptions, GrepProcessFactory } from './grep-workspace';
+
+export const grepProcessModule = new ContainerModule(bind => {
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-process-module.ts]: let's rename this

File: packages/grep-workspace/src/node/grep-process-module.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    bind(GrepWS).toSelf().inTransientScope();
+    bind(GrepProcessFactory).toFactory(ctx =>
+        (options: GrepProcessOptions) => {
+            const child = new Container({ defaultScope: 'Singleton' });
+            child.parent = ctx.container;
+
+            const logger = ctx.container.get<ILogger>(ILogger);
+            const loggerChild = logger.child({ 'module': 'process' });
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-process-module.ts]: This is no longer the 'process' module so change the logging

File: packages/grep-workspace/src/node/grep-workspace.spec.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import { testContainer } from './inversify.spec-config';
+import { GrepProcessFactory } from './grep-workspace';
+
+const expect = chai.expect;
+
+describe("grep-workspace", () => {
+    const grepProcessFactory = testContainer.get<GrepProcessFactory>(GrepProcessFactory);
+
+    it('test pipe stream', function () {
+        const grepWorkSpace = grepProcessFactory({ command: process.execPath, 'args': ['--version'] });
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-workspace.spec.ts]: what are you testing here ?

File: packages/grep-workspace/src/node/grep-workspace.spec.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        const outStream = new stream.PassThrough();
+
+        const p = new Promise<String>((resolve, reject) => {
+            outStream.on('data', data => {
+                console.log("data received is : " + data.toString());
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-workspace.spec.ts]: we should not leave that here

File: packages/grep-workspace/src/node/grep-workspace.spec.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            only 'exit' is sent on the terminal process.  */
+            grepWorkSpace.onExit(() => {
+                grepWorkSpace.dispose();
+            });
+        });
+
+        grepWorkSpace.output.pipe(outStream);
+
+        /* Avoid using equal since terminal characters can be inserted at the end.  */
+        //        return expect(p).to.eventually.have.string(process.version);
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-workspace.spec.ts]: remove comment

File: packages/grep-workspace/src/node/grep-workspace.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable, inject } from 'inversify';
+import * as process from 'process';
+import { ILogger } from '@theia/core/lib/common/logger';
+import { TerminalProcess, TerminalProcessOptions } from '@theia/process/lib/node/terminal-process';
+import { ProcessManager } from '@theia/process/lib/node';
+
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-workspace.ts]: empty line

File: packages/grep-workspace/src/node/grep-workspace.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+export const GrepProcessOptions = Symbol("GrepProcessOptions");
+export interface GrepProcessOptions {
+    command: string,
+    args?: string[],
+    options?: object
+}
+export const GrepProcessFactory = Symbol("GrepProcessFactory");
+export type GrepProcessFactory = (options: GrepProcessOptions) => GrepWS;
+
+
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-workspace.ts]: just 1 empty line is ok

File: packages/grep-workspace/src/node/grep-workspace.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+import { rgPath } from 'vscode-ripgrep';
+
+
+export const GrepProcessOptions = Symbol("GrepProcessOptions");
+export interface GrepProcessOptions {
+    command: string,
+    args?: string[],
+    options?: object
+}
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-workspace.ts]: Does these options make sense for ripgrep ?

I don't think it needs command since we know what it is ...
args is OK I guess
does it need options?

File: packages/grep-workspace/src/node/grep-process-module.ts (hunk 5 above)
Comment (hexa00) [reply packages/grep-workspace/src/node/grep-process-module.ts]: Also please avoid abreviations

File: packages/grep-workspace/src/node/grep-workspace.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+import { injectable, inject } from 'inversify';
+import * as process from 'process';
+import { ILogger } from '@theia/core/lib/common/logger';
+import { TerminalProcess, TerminalProcessOptions } from '@theia/process/lib/node/terminal-process';
+import { ProcessManager } from '@theia/process/lib/node';
+
+import { rgPath } from 'vscode-ripgrep';
+
+
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-workspace.ts]: 1 empty line only

File: packages/grep-workspace/src/node/grep-workspace.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+}
+export const GrepProcessFactory = Symbol("GrepProcessFactory");
+export type GrepProcessFactory = (options: GrepProcessOptions) => GrepWS;
+
+
+@injectable()
+export class GrepWS extends TerminalProcess {
+
+    protected static defaultCols = 80;
+    protected static defaultRows = 24;
```
Comment (hexa00) [inline packages/grep-workspace/src/node/grep-workspace.ts]: why this here?

File: packages/grep-workspace/src/package.spec.ts [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+/* note: this bogus test file is required so that
+   we are able to run mocha unit tests on this
+   package, without having any actual unit tests in it.
+   This way a coverage report will be generated,
+   showing 0% coverage, instead of no report.
+   This file can be removed once we have real unit
+   tests in place. */
+
```
Comment (hexa00) [inline packages/grep-workspace/src/package.spec.ts]: we have a test so we don't need that file

File: packages/grep-workspace/src/typings/vscode-ripgrep.d.ts [hunk 16]
```diff
@@ -0,0 +1,3 @@
+declare module 'vscode-ripgrep' {
+	export const rgPath: string;
+}
```
Comment (hexa00) [inline packages/grep-workspace/src/typings/vscode-ripgrep.d.ts]: is that needed ?

File: packages/tsconfig.json [hunk 17]
```diff
@@ -57,12 +57,12 @@
             "@theia/go/lib/*": [
                 "go/src/*"
             ],
-            "@theia/process/lib*": [
-                "process/src/*"
```
Comment (hexa00) [inline packages/tsconfig.json]: why are you replacing process here ???

File: packages/grep-workspace/src/typings/vscode-ripgrep.d.ts (hunk 16 above)
Comment (lmcgupe) [reply packages/grep-workspace/src/typings/vscode-ripgrep.d.ts]: Moved locally.

File: packages/grep-workspace/src/node/grep-workspace.spec.ts (hunk 7 above)
Comment (lmcgupe) [reply packages/grep-workspace/src/node/grep-workspace.spec.ts]: removed the command parameter.

File: packages/grep-workspace/src/node/grep-workspace.spec.ts (hunk 8 above)
Comment (lmcgupe) [reply packages/grep-workspace/src/node/grep-workspace.spec.ts]: i want to keep it for now to see the output of the command. I added a note to remove it in the end.

**Comment by hexa00** [comment]
We need to figure-out how to download or package ripgrep differently since it fails the build often because of https://github.com/Microsoft/vscode/issues/28434

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.ts [hunk 18]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.ts]: the copy right is there twice ? 

You can remove the typefox one...

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.ts [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+import { ILogger } from '@theia/core/lib/common/logger';
+import { TerminalProcess, TerminalProcessOptions } from '@theia/process/lib/node/terminal-process';
+import { ProcessManager } from '@theia/process/lib/node';
+import { rgPath } from 'vscode-ripgrep';
+
+export const RipGrepProcessOptions = Symbol("GrepProcessOptions");
+export interface RipGrepProcessOptions {
+    args?: string[],
+}
+export const RipGrepProcessFactory = Symbol("GrepProcessFactory");
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.ts]: Symbol is wrong

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.ts [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+import { injectable, inject } from 'inversify';
+import * as process from 'process';
+import { ILogger } from '@theia/core/lib/common/logger';
+import { TerminalProcess, TerminalProcessOptions } from '@theia/process/lib/node/terminal-process';
+import { ProcessManager } from '@theia/process/lib/node';
+import { rgPath } from 'vscode-ripgrep';
+
+export const RipGrepProcessOptions = Symbol("GrepProcessOptions");
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.ts]: symbol is wrong

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.ts [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+@injectable()
+export class RipGrepWorkSpace extends TerminalProcess {
+
+    constructor(
+        @inject(RipGrepProcessOptions) options: RipGrepProcessOptions,
+        @inject(ProcessManager) processManager: ProcessManager,
+        @inject(ILogger) logger: ILogger
+    ) {
+
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.ts]: empty line

File: packages/tsconfig.json (hunk 17 above)
Comment (hexa00) [reply packages/tsconfig.json]: still removed ?

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.ts [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+        super(<TerminalProcessOptions>{
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.ts]: I think you can remove the cast...

I should remove it from shell too

**Comment by hexa00** [comment]
We should add more functionality and test it in the class like search pattern in directory

**Comment by hexa00** [comment]
See: https://github.com/roblourens/vscode-ripgrep/blob/master/dist/postinstall.js
```
 const opts = {
        platform: os.platform(),
        version: '0.5.1-patch.0',
        token: process.env['GITHUB_TOKEN']
};
```
So we should put the token in the GITHUB_TOKEN env var and remove the echo to npmrc since it's not npm that fetches from github

File: packages/ripgrep-workspace/src/node/inversify.spec-config.ts [hunk 23]
```diff
@@ -0,0 +1,16 @@
+/*
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/inversify.spec-config.ts]: should be in node/test/

File: packages/ripgrep-workspace/src/node/ripgrep-process-module.ts [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+import { RipGrepWorkSpace, RipGrepProcessOptions, RipGrepProcessFactory } from './ripgrep-workspace';
+
+export const ripgrepProcessModule = new ContainerModule(bind => {
+    bind(RipGrepWorkSpace).toSelf().inTransientScope();
+    bind(RipGrepProcessFactory).toFactory(ctx =>
+        (options: RipGrepProcessOptions) => {
+            const child = new Container({ defaultScope: 'Singleton' });
+            child.parent = ctx.container;
+            const logger = ctx.container.get<ILogger>(ILogger);
+            const loggerChild = logger.child({ 'module': 'ripgrepProcess' });
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-process-module.ts]: we should do this with 1 logger as a singleton per package see terminal-backend-module.ts

File: packages/ripgrep-workspace/src/node/ripgrep-process-module.ts [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+            child.bind(RipGrepProcessOptions).toConstantValue(options);
+            child.bind(ILogger).toConstantValue(loggerChild);
+            return child.get(RipGrepWorkSpace);
+        }
+    );
+
+    bind(ProcessManager).toSelf().inSingletonScope();
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-process-module.ts]: Why is this here ??

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+        const p = new Promise<String>((resolve, reject) => {
+            // TODO: for debugging only, remove in final version
+            outStream.on('data', data => {
+                console.log("data received is : " + data.toString());
+            });
+        });
+
+        ripgrepWorkSpace.output.pipe(outStream);
+
+        return expect(p).to.not.equal(null);
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts]: This will always be true, what is this testing?

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+import { testContainer } from './inversify.spec-config';
+import { RipGrepProcessFactory } from './ripgrep-workspace';
+
+const expect = chai.expect;
+
+describe("ripgrep-workspace", () => {
+    const ripgrepProcessFactory = testContainer.get<RipGrepProcessFactory>(RipGrepProcessFactory);
+
+    it('test basic option', function () {
+        const ripgrepWorkSpace = ripgrepProcessFactory({ 'args': ['--version'] });
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts]: should be in beforeEach if you do that in multiple tests

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+        const p = new Promise<String>((resolve, reject) => {
+            // TODO: for debugging only, remove in final version
+            outStream.on('data', data => {
+                console.log("grep data received is : " + data.toString());
+            });
+        });
+
+        ripgrepWorkSpace.output.pipe(outStream);
+
+        return expect(p).to.not.equal(null);
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts]: ditto what is this testing?

File: packages/ripgrep-workspace/src/node/ripgrep-process-module.ts (hunk 25 above)
Comment (lmcgupe) [reply packages/ripgrep-workspace/src/node/ripgrep-process-module.ts]: it needs ProcessManager.

Comment (hexa00) [reply packages/ripgrep-workspace/src/node/ripgrep-process-module.ts]: But how does binding it here satisfy that need ?

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+const ripgrepProcessFactory = testContainer.get<RipGrepProcessFactory>(RipGrepProcessFactory);
+
+describe("ripgrep-workspace", () => {
+    it('test basic option', function () {
+        ripgrepWorkSpace = ripgrepProcessFactory({ 'args': ['--version'] });
+        const outStream = new stream.PassThrough();
+
+        outStream.on('data', data => {
+            expect(data.toString()).to.include("ripgrep");
+        });
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts]: this test doesn't work please read about the mocha done() callback

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+    });
+    it('test basic grep', function () {
+        ripgrepWorkSpace = ripgrepProcessFactory({ 'args': ['theia *'] });
+        const outStream = new stream.PassThrough();
+
+        outStream.on('data', data => {
+            expect(data.toString()).to.include("theia");
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts]: ditto

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.ts [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+
+import { injectable, inject } from 'inversify';
+import * as process from 'process';
+import { ILogger } from '@theia/core/lib/common/logger';
+import { TerminalProcess } from '@theia/process/lib/node/terminal-process';
+import { ProcessManager } from '@theia/process/lib/node';
+import { rgPath } from 'vscode-ripgrep';
+let os = require('os');
+let commandLine = rgPath;
+if (os.platform().startsWith('win')) {
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.ts]: check for isWindows in theia's os.ts and use that

File: packages/ripgrep-workspace/extension.package.json [hunk 32]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "@theia/ripgrep-workspace",
+  "version": "0.1.1",
+  "description": "Theia - RipGrep Workspace Extension",
+  "dependencies": {
+    "@theia/core": "^0.1.1",
+    "@theia/process": "^0.1.1",
+    "@theia/languages": "^0.1.1",
+    "@theia/editor": "^0.1.1",
+    "@theia/monaco": "^0.1.1",
```
Comment (hexa00) [inline packages/ripgrep-workspace/extension.package.json]: do we need all this ? monaco editor languages ?? I don't think so

File: packages/ripgrep-workspace/src/node/ripgrep-process-module.ts (hunk 25 above)
Comment (lmcgupe) [reply packages/ripgrep-workspace/src/node/ripgrep-process-module.ts]: not sure. where else should it go ?

Comment (hexa00) [reply packages/ripgrep-workspace/src/node/ripgrep-process-module.ts]: Well were is it already? What does it do ?

File: packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+            const p = new Promise<String>((resolve, reject) => {
+                outStream.on('data', data => {
+                    resolve(data.toString());
+                });
+            });
+            ripgrepWorkSpace.output.pipe(outStream);
+            p.then((data) => {
+                expect(data).to.include("theia");
+                done();
+            })
```
Comment (hexa00) [inline packages/ripgrep-workspace/src/node/ripgrep-workspace.spec.ts]: You just actually write that as
`return expect(p).to.eventually.include("theia")`
same thing for all the tests

File: packages/ripgrep-workspace/src/common/search-workspace-protocol.ts [hunk 34]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+import { JsonRpcServer } from '@theia/core/lib/common/messaging/proxy-factory';
+
+export const ISearchWorkSpaceServer = Symbol('ISearchWorkSpaceServer');
+
+export const searchPath = '/services/search';
+
+export interface ISearchWorkSpaceServer extends JsonRpcServer<ISearchWorkSpaceClient> {
+    query(regExp: string): Promise<string>;
```
Comment (svenefftinge) [inline packages/ripgrep-workspace/src/common/search-workspace-protocol.ts]: Should I not be able to specify the path under which the search should happen?

Comment (hexa00) [reply packages/ripgrep-workspace/src/common/search-workspace-protocol.ts]: This is still very WIP @lmcbout is just trying to make a frontend/backend service.
So this is just for testing that...

**Comment by svenefftinge** [comment]
Closing this in favor of #1046
